econml/solutions/causal_analysis/_causal_analysis.py

```diff
@@ -0,0 +1,1071 @@
+
+    # Create a summary combining all results into a single output; this is used
+    # by the various causal_effect and causal_effect_dict methods to generate either a dataframe
+    # or a dictionary, respectively, based on the summary function passed into this method
+    def _summarize(self, *, summary, get_inference, props, expand_arr, drop_sample):
+
+        assert hasattr(self, "_results"), "This object has not been fit, so cannot get results"
+
+        # ensure array has shape (m,y,t)
+        def ensure_proper_dims(arr):
+            if expand_arr:
+                # population summary is missing sample dimension; add it for consistency
+                arr = np.expand_dims(arr, 0)
+            if self._vec_y:
+                # outcome dimension is missing; add it for consistency
+                arr = np.expand_dims(arr, axis=1)
+            assert 2 <= arr.ndim <= 3
+            # add singleton treatment dimension if missing
+            return arr if arr.ndim == 3 else np.expand_dims(arr, axis=2)
+
+        # each attr has dimension (m,y) or (m,y,t)
+        def coalesce(attr):
+            """Join together the arrays for each feature"""
+            attr = self._make_accessor(attr)
+            # concatenate along treatment dimension
+            arr = np.concatenate([ensure_proper_dims(attr(get_inference(res)))
+                                  for res in self._results], axis=2)
+
+            # for dictionary representation, want to remove unneeded sample dimension
+            # in cohort and global results
+            if drop_sample:
+                arr = np.squeeze(arr, 0)
+
+            return arr
+
+        return summary([(key, coalesce(val)) for key, val in props])
+
+    def _pandas_summary(self, get_inference, props, n,
+                        expand_arr=False):
+        """
+        Summarizes results into a dataframe.
+
+        Parameters
+        ----------
+        get_inference : lambda
+            Method to get the relevant inference results from each result object
+        props : list of (string, string or lambda)
+            Set of column names and ways to get the corresponding values from the inference object
+        n : int
+            The number of samples in the dataset
+        expand_arr : boolean, default False
+            Whether to add an initial sample dimension to the result arrays
+        """
+        def make_dataframe(props):
+
+            to_include = OrderedDict([(key, value.reshape(-1))
+                                      for key, value in props])
+
+            # TODO: enrich outcome logic for multi-class classification when that is supported
+            index = pd.MultiIndex.from_tuples([(i, outcome, res.feature_name, f"{lvl}v{res.feature_baseline}"
+                                                if res.feature_baseline is not None
+                                                else lvl)
+                                               for i in range(n)
```

Comment:
My feeling is that the large extra overhead in local causal effect is stemming from the for i in range(n) loop which is a terrible for loop of 5k samples for instance in native python. We should find a way to avoid a pytrhon for loop over the samples if possible. Also we should first make sure that it is indeed coming from this.

Reply:
Let's add performance improvement on this in the TODOs for next release